### DISCUSSION TITLE
ideaspark ESP32 ST7789 170x320

### DISCUSSION THREAD

**Discussion by TheChatty** [arendst/Tasmota#22669] [Support request]
I've got this ESP:
[![image](https://github.com/user-attachments/assets/cf60d670-6306-4964-8d80-58f4c80c2957)](https://www.aliexpress.com/item/1005007181435830.html)

[![image](https://github.com/user-attachments/assets/15ec00ca-790e-4a54-9f38-f082c5b83acd)](https://ae-pic-a1.aliexpress-media.com/kf/S592c0c0eb5ec42e9ad0cdcaa0eb7b03aA.png)

I'm using [tasmota32-lvgl.factory.bin](https://ota.tasmota.com/tasmota32/release/tasmota32-lvgl.factory.bin) and [ST7789_display.ini](https://github.com/arendst/Tasmota/blob/development/tasmota/displaydesc/ST7789_display.ini) except first line changed to `:H,ST7789,170,320,16,SPI,1,*,*,*,*,*,*,*,40`.

With this setup I can switch the display on/off including backlight and have the following boot up screen:
![IMG_4569](https://github.com/user-attachments/assets/32658397-e3f8-42ea-821d-b53f461f16d1)

The [information about display.ini](https://tasmota.github.io/docs/Universal-Display-Driver/#descriptor-file) didn't help me. So please help in the necessary adjustments.

A user from aliexpress successfully setup ESPHome with [this config](https://ae-pic-a1.aliexpress-media.com/kf/A86d389c0cd664c3c86eb64b01128d06f8.png_.webp) - it includes an offset_height of 35. Is that a hint?

There is also a small microswitch in the upper left corner on the back side (in relation to my photo) connected to GPIO4. If I press it (LOW -> HIGH), the display goes blank and stays that way when using the on/off toggle.

**Comment by sfromis** [comment]
The meaning of entries in `display.ini` are described here:
https://tasmota.github.io/docs/Universal-Display-Driver/#descriptor-file
An obvious difference is the resolution in the first line.

**Reply by TheChatty** [reply]
As I said right below the pinout picture: the first line was changed by me already.

**Reply by s-hadinger** [reply]
There are a few changes needed.

The screen seems to be inverted, so `:i,21,20` should invert again

The device does not seem to have a touch screen, I removed the section.

Unfortunately I don't know how to set an offset for ST7789

```
:H,ST7789,170,320,16,SPI,1,*,*,*,*,*,*,*,40
:S,2,1,3,0,80,30
:I
01,A0
11,A0
3A,81,55
36,81,00
21,80
13,80
29,A0
:o,28
:O,29
:A,2A,2B,2C
:R,36
:0,C0,00,50,00
:1,A0,50,00,01
:2,00,00,00,02
:3,60,00,00,03
:i,21,20
#
```

**Reply by s-hadinger** [reply]
Are you sure it's a ST7789? The Esphome mentions also ILI9xxx

**Comment by TheChatty** [comment]
The following observations are described when holding the module in portrait (usb connector on lower side):

With **DisplayRotate 0** and
**x size 170**: 35 pixels undefined on the left and 35 pixels on the right outside of screen
**y size 320**: 80 pixels outside of screen on the top and 80 pixels undefined on the bottom

With **DisplayRotate 2** and
**x size 170**: ~35 pixels on the left outside of screen and 35 pixels undefined on the right
**y size 320**: looking alright

With **DisplayRotate 0** and
**x size 205**: good on the left and 35 pixels on the right outside of screen
**y size 320**: 80 pixels outside of screen on the top and 80 pixels undefined on the bottom

With **DisplayRotate 2** and
**x size 205**: 35 pixels on the left outside of screen /  **good** on the right
**y size 320**: **all good**

![image](https://github.com/user-attachments/assets/07829150-2a55-46da-8d58-03d86f2a1a90)

@s-hadinger: Your updated display.ini works as expected.

The component from ESPHome that drives most displays is called [ILI9xxx](https://esphome.io/components/display/ili9xxx#ili9xxx). This driver supports "removing" those outside areas (see `offset_width`).

So maybe uDisplay needs this setting as well?

Looking closer at the ESPHome configuration I see that the other user also set `invert_colors: true`, `pixel_mode: 18bit` and `data_rate: 80MHz`.

80Mhz is also working for me with uDisplay.

@Edit: I just tested ESPHome and it works "out of the box" (320x170, 80 MHz, colors right, 18bit). The display also does not recover once I pressed GPIO4 (LCD Reset).

**Reply by s-hadinger** [reply]
Oh, I overlooked the `display.ini`. It's actually much simpler to configure.

For `DisplayRotate 0`

Get back to 170x320
```
:H,ST7789,170,320,16,SPI,1,*,*,*,*,*,*,*,40
```

And set the offset here (0x23 is 35):
```
:0,C0,23,00,00
```

I will let you find the right values for other rotations

**Comment by TheChatty** [comment]
Thank you. All rotations are working now. I created a [PR](https://github.com/arendst/Tasmota/pull/22678).

Only things left:
* How to recover display after GPIO4?
* How to utilize 18bit depth?

**Reply by s-hadinger** [reply]
Why would you like to reset the display with GPIO4? It is normally done at boot.

18 bits depth is not supported by LVGL, only 16 bits

**Reply by TheChatty** [reply]
I don't need to reset the display but I'd like an already attached usable button. GPIO0 has such a button on the front.

**Reply by s-hadinger** [reply]
I understand that gpio4 is connected to display reset. So not usable

**Reply by TheChatty** [reply]
I can't say why it wasn't working before but **now it works**. If I press the GPIO4 button display turns dark but then I issue `displayreinit` and after redrawing everything looks good.

So this **button is usable** but it has to be supported by a rule which reinits the display. Is there actually a simple `lv.scr_act().redraw()` function?

**Reply by s-hadinger** [reply]
The API reference is here: https://tasmota.github.io/docs/LVGL_API_Reference/

The closest should be `lv.scr_act().invalidate()`

**Reply by TheChatty** [reply]
Cool, thanks! That works.
